### PR TITLE
tikv: eliminate unnecessary pessimistic rollback

### PR DESCRIPTION
**Pull request by youjiali1995**
Signed-off-by: youjiali1995 <zlwgx1023@gmail.com>

### What problem does this PR solve? 
When acquiring pessimistic locks fails, we do async pessimistic rollback to clean up all locks even some of them are not locked. These rollback operations will conflict with the following lock operations which may reduce performance.

### What is changed and how it works?
If there is only 1 key and lock failed, no need to do pessimistic rollback.

sysbench oltp_write_only, tables: 32*100, threads: 128

Before
```
SQL statistics:
    queries performed:
        read:                            0
        write:                           28523
        other:                           14900
        total:                           43423
    transactions:                        5859   (47.05 per sec.)
    queries:                             43423  (348.71 per sec.)
    ignored errors:                      3182   (25.55 per sec.)
    reconnects:                          0      (0.00 per sec.)

General statistics:
    total time:                          124.5217s
    total number of events:              5859

Latency (ms):
         min:                                    1.77
         avg:                                 2694.71
         max:                                33254.30
         95th percentile:                     9624.59
         sum:                             15788288.77
```

This PR
```
SQL statistics:
    queries performed:
        read:                            0
        write:                           342892
        other:                           179245
        total:                           522137
    transactions:                        70701  (235.34 per sec.)
    queries:                             522137 (1737.99 per sec.)
    ignored errors:                      37843  (125.96 per sec.)
    reconnects:                          0      (0.00 per sec.)

General statistics:
    total time:                          300.4248s
    total number of events:              70701

Latency (ms):
         min:                                    1.88
         avg:                                  543.52
         max:                                 8555.44
         95th percentile:                     1561.52
         sum:                             38427743.18
```

### Check List 

Tests 
 
- No code

Code changes

Side effects

Related changes

 - Need to cherry-pick to the release branch

Release note

 - Improve the performance of the pessimistic transaction in some situations with serious conflicts.

### CODE REVIEW

File: store/tikv/txn.go [hunk 1]
```diff
@@ -428,13 +428,16 @@ func (txn *tikvTxn) LockKeys(ctx context.Context, forUpdateTS uint64, keysInput
...
 			}
-			wg := txn.asyncPessimisticRollback(ctx, keys)
-			if dl, ok := errors.Cause(err).(*ErrDeadlock); ok && hashInKeys(dl.DeadlockKeyHash, keys) {
-				dl.IsRetryable = true
-				// Wait for the pessimistic rollback to finish before we retry the statement.
-				wg.Wait()
-				// Sleep a little, wait for the other transaction that blocked by this transaction to acquire the lock.
-				time.Sleep(time.Millisecond * 5)
+			// If there is only 1 key and lock failed, no need to do pessimistic rollback.
+			if len(keys) > 1 {
```
Comment (coocood) [inline store/tikv/txn.go]: We need to make sure the err indicates the lock must not be locked to skip pessimistic rollback.

**Comment by youjiali1995** [comment]
@coocood PTAL again. Thanks.

**Comment by coocood** [comment]
LGTM

**Comment by coocood** [comment]
@tiancaiamao PTAL

File: store/tikv/txn.go [hunk 2]
```diff
@@ -428,13 +429,17 @@ func (txn *tikvTxn) LockKeys(ctx context.Context, forUpdateTS uint64, keysInput
...
 				txn.us.DeleteKeyExistErrInfo(key)
 			}
-			wg := txn.asyncPessimisticRollback(ctx, keys)
-			if dl, ok := errors.Cause(err).(*ErrDeadlock); ok && hashInKeys(dl.DeadlockKeyHash, keys) {
-				dl.IsRetryable = true
-				// Wait for the pessimistic rollback to finish before we retry the statement.
-				wg.Wait()
-				// Sleep a little, wait for the other transaction that blocked by this transaction to acquire the lock.
-				time.Sleep(time.Millisecond * 5)
+			keyMayBeLocked := terror.ErrorNotEqual(kv.ErrWriteConflict, err) && terror.ErrorNotEqual(kv.ErrKeyExists, err)
```
Comment (MyonKeminta) [inline store/tikv/txn.go]: I'm afraid if we add more error types or changes our usage of these error types in the future, this may fail.

Comment (youjiali1995) [reply store/tikv/txn.go]: That is why it use `ErrorNotEqual` here. As long as we make sure `ErrWriteConflict` and `ErrKeyExists` won't lock the key.

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**